apps/fipsinstall.c

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include "apps.h"
+#include "progs.h"
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/provider.h>
+#include <openssl/params.h>
+
+#define BUFSIZE 4096
+#define DEFAULT_MAC_NAME "HMAC"
+#define DEFAULT_FIPS_SECTION "fips_check_section"
+
+/* Configuration file values */
+#define VERSION_KEY  "version"
+#define VERSION_VAL  "1"
+#define INSTALL_STATUS_VAL "INSTALL_SELF_TEST_KATS_RUN"
+
+typedef enum OPTION_choice {
+    OPT_ERR = -1, OPT_EOF = 0, OPT_HELP,
+    OPT_IN, OPT_OUT,
+    OPT_PROV_NAME, OPT_MODULE, OPT_SECTION_NAME,
+    OPT_MACOPT, OPT_VERIFY, OPT_MAC
+} OPTION_CHOICE;
+
+const OPTIONS fipsinstall_options[] = {
+    {"help", OPT_HELP, '-', "Display this summary"},
+    {OPT_MORE_STR, 0, 0, "e.g: openssl fipsinstall -name fips -section "
+     "fipsinstall -out fips.conf -module ./fips.so -mac HMAC "
+     "-macopt digest:SHA256 -macopt hexkey:00"},
+    {"verify", OPT_VERIFY, '-', "Verification mode, i.e verify a config file "
+     "instead of generating one"},
+    {"in", OPT_IN, 's', "Input config file, used when verifying"},
+    {"out", OPT_OUT, 's', "Output config file, used when generating"},
+    {"module", OPT_MODULE, '>', "File name of the provider module"},
```

Comment:
Those three should be:

``` C
    {"in", OPT_IN, '<', "Input config file, used when verifying"},
    {"out", OPT_OUT, '>', "Output config file, used when generating"},
    {"module", OPT_MODULE, '<', "File name of the provider module"},
```

The difference is in the third element, indicating if it's an input file or an output file.